File: code/universe.py
from math import sqrt, exp, atan, pi, cos, copysign, sin, log
from random import random

from bhtree import BHTree
from body import Body
from quad import Quad


class Universe:
    """A container for orbital dynamics simulations"""
    bodies: [Body] = []
    solar_mass = 1.98892e30
    radius = 1e18  # the radius of the universe!!

    @property
    def n(self):
        return len(self.bodies)

    def __init__(self):
        pass

    def calculate(self):
        self.add_forces(self.n)

    def circular_velocity(self, rx: float, ry: float) -> float:
        """
        the bodies are initialized in circular orbits around the central mass.
        This is just some physics to do that
        """
        r2 = sqrt(rx*rx + ry*ry)
        numerator = 6.67e-11 * 1e6 * self.solar_mass
        return sqrt(numerator/r2)

    def start_the_bodies(self, n: int):
        """
        Initialize N bodies with random positions and circular velocities
        """

        # Put something heavy into the center
        self.bodies.append(Body(0, 0, 0, 0, 1e6 * self.solar_mass))

        for i in range(n):
            px = 1e18 * exp(-1.8) * (.5-random())
            py = 1e18 * exp(-1.8) * (.5-random())
            magv = self.circular_velocity(px, py)

            absangle = atan(abs(py / px))
            thetav = pi / 2 - absangle
            phiv = random() * pi
            # https://finnaarupnielsen.wordpress.com/2011/05/18/where-is-the-sign-function-in-python/
            vx = -1 * copysign(1, py) * cos(thetav) * magv
            vy = copysign(1, px) * sin(thetav) * magv

            # Orient a random 2D circular orbit
            if random() <= 0.5:
                vx = -vx
                vy = -vy

            mass = random() * self.solar_mass * 10 + 1e20
            self.bodies.append(Body(px, py, vx, vy, mass))

    def add_forces(self, n):
        """Override this method in your subclass"""
        raise NotImplementedError

    def exp(self, lmbda):
        """
        A function to return an exponential distribution for position
        """
        return -log(1 - random()) / lmbda


class BruteForceUniverse(Universe):
    """
    Naive implementation of orbital dynamics
    """
    def add_forces(self, n):
        for body in self.bodies:
            body.reset_force()

            # uh, oh a loop inside a loop
            # This'll get us n^2 complexity
            for other_body in self.bodies:
                if other_body != body:
                    body.add_force(other_body)

        # Update the timestamps
        for body in self.bodies:
            body.update(1e11)


class BarnesHutUniverse(Universe):
    """
    Universe that uses a Barnes-Hut tree to calculte orbital dynamics
    """
    quad = Quad(0, 0, 2*1e18)

    def add_forces(self, n):
        the_tree = BHTree(self.quad)

        # If the body is still on the screen, add it to the tree
        for body in self.bodies:
            if body.is_in(self.quad):
                the_tree.insert(body)

        # Now, use out methods in BHTree to update the forces,
        # traveling recursively through the tree
        for body in self.bodies:
            body.reset_force()
            if body.is_in(self.quad):
                the_tree.update_force(body)
                # Calculate the new positions on a time step dt (1e11 here)
                body.update(1e11)

File: NBody/Simulation/Universe.swift
//
//  Universe.swift
//  NBody
//
//  Created by Daniel Jilg on 14.04.18.
//  Copyright © 2018 breakthesystem. All rights reserved.
//

import AppKit
import Foundation

/// A container for simulation of orbital dynamics
class Universe {
    // MARK: Properties
    public private(set) var bodies = [Body]()

    // MARK: - Configuration
    public let solarMass = 1.988435e30
    public let smallestPlanetMass = 3.30104e23
    public let largestPlanetMass = 1.89813e27
    public let radius = 1e12
    public let parentBodyColor = #colorLiteral(red: 0.9529411793, green: 0.6862745285, blue: 0.1333333403, alpha: 1)

    /// Initialize N bodies with random positions and circular-ish velocities
    init(numberOfBodies: Int) {
        // Put something heavy into the center
        let sun = Planetoid(
            position: Position(x: 0, y: 0),
            direction: Vector(x: 0, y: 0),
            name: "Sün",
            tickNumber: 0,
            mass: solarMass
        )
        sun.color = parentBodyColor
        self.bodies.append(sun)

        // Put some bodies around it
        for i in 1...numberOfBodies {
            // let position = Position(x: (drand48() * 2 - 1) * radius, y: (drand48() * 2 - 1) * radius)
            let freeSpace = 0.5

            let positionX = drand48() * radius * (1-freeSpace) + radius * freeSpace * (drand48() > 0.5 ? 1 : -1)
            let positionY = drand48() * radius * (1-freeSpace) + radius * freeSpace * (drand48() > 0.5 ? 1 : -1)
            let position = Position(x: positionX , y: positionY)
            let velocity = circularVelocity(for: position)
            let mass = smallestPlanetMass + ((largestPlanetMass - smallestPlanetMass) * drand48())
            bodies.append(Planetoid(position: position, direction: velocity, name: "Body \(i)", tickNumber: 0, mass: mass))
        }
    }

    func update(elapsedTime: Double) {
        fatalError("Not Implemented")
    }


    // MARK: Helper Methods
    /// Return the orbital velocity an orbiting body should have at a given position to stay in a stable orbit
    ///
    /// This function ignores the orbiting body's mass, so it's not very accurate for high-mass bodies
    /// But that's ok since we don't want perfect boring orbits anyway.
    public func circularVelocity(for position: Position) -> Vector {
        // Calculate orbital velocity
        let standardGravitationalParameter = G * solarMass
        let orbitalRadius = position.distance(to: Position(x: 0, y: 0))
        var orbitalVelocity = sqrt(standardGravitationalParameter / orbitalRadius)

        // Reverse velocity for a small percentage of bodies
        if drand48() < 0.9 {
            orbitalVelocity = -orbitalVelocity
        }

        // calculate angle at given position
        let absoluteAngle = atan(abs(position.y / position.x))
        let thetaV = Double.pi / 2 - absoluteAngle

        // calculate velocity vector at given angle
        let vectorX = (position.y > 0 ? -1 : 1) * cos(thetaV) * orbitalVelocity
        let vectorY = (position.x > 0 ? 1 : -1) * sin(thetaV) * orbitalVelocity
        return Vector(x: vectorX, y: vectorY)
    }
}

class BruteForceUniverse: Universe {
    override func update(elapsedTime: Double) {
        for body in bodies where body is Planetoid {
            body.resetForce()

            // uh, oh a loop inside a loop
            // This'll get us n^2 complexity
            for otherBody in bodies where body is Planetoid {
                if otherBody !== body {
                    (body as! Planetoid).accelerate(with: (otherBody as! Planetoid))
                }
            }
        }

        for body in bodies {
            body.update(timeSteps: elapsedTime)
        }
    }
}
